jung-algorithms/src/main/java/edu/uci/ics/jung/algorithms/util/DiscreteDistribution.java

```diff
@@ -165,43 +164,36 @@ public static void normalize(double[] counts, double alpha) {
    * @see #mean(double[][])
    * @param distributions the distributions whose mean is to be calculated
    * @return the mean of the distributions
+   * @throws IllegalArgumentException if the outer distributions array is empty, or if any of the
+   *     inner distribution arrays are empty, or if not all of the inner distribution arrays are the
+   *     same length
    */
   public static double[] mean(Collection<double[]> distributions) {
-    Preconditions.checkArgument(
-        !distributions.isEmpty(), "Distribution collection must be non-empty");
-    // TODO: Consider checking that the inner arrays of `distributions` are non-empty
-    // TODO: Consider using com.google.common.math.Stats.meanOf
-    Iterator<double[]> iter = distributions.iterator();
-    double[] first = iter.next();
-    double[][] dArray = new double[distributions.size()][first.length];
-    dArray[0] = first;
-    for (int i = 1; i < dArray.length; i++) {
-      dArray[i] = iter.next();
-    }
-
-    return mean(dArray);
+    checkArgument(!distributions.isEmpty(), "Distribution collection must be non-empty");
```

Comment:
As much as a like the clean and modern look of streams, I suspect that the time required to create new Objects is > 0 so their inclusion in algorithmic code can only decrease performance.

Reply:
Hi @tomnelson. I realise that I've said this a few times in the past, but I'm not convinced yet that using streams is such a problem for performance.

Whereas I agree that streams can and usually do increase the time needed to complete operations compared to for-loops, I don't think it's that much extra time for a good deal of things. Furthermore, arguably performance-sensitive software like error-prone, Bazel and the OpenJDK standard library use streams (or even Guava's functional constructs) in some cases, and JUnit 5 uses them extensively, so it's not obvious to me that we should immediately reject the use of streams in JUNG.

Do you have any numbers, like from a JMH benchmark or a profiler run on a real-life application, that show that `DiscreteDistribution.mean()` would benefit greatly from an imperative implementation?

Reply:
"Imperative code should be your default, your first choice as of Java 7. You should not use functional idioms unless you are absolutely sure of one of the following:

Use of functional idioms will result in net savings of lines of code for your entire project. In the example above, the "functional" version used 11 lines, the imperative version 6. Moving the definition of a function to another file, or a constant, does not help.
For efficiency, you need a lazily computed view of the transformed collection and cannot settle for an explicitly computed collection. Additionally, you have read and reread Effective Java, item 55, and besides following those instructions, you have actually done benchmarking to prove that this version is faster, and can cite numbers to prove it."

From https://github.com/google/guava/wiki/FunctionalExplained . I couldn't say it better myself.

Reply:
@phrocker Please note that that part of the `FunctionalExplained` doc was written before Java 8 came out.  The top of the doc, however, contextualizes it:

_Java 8 includes the java.util.function and java.util.stream packages, which supersede Guava's functional programming classes for projects at that language level._

_While Guava's functional utilities are usable on Java versions prior to Java 8, functional programming without Java 8 requires awkward and verbose use of anonymous classes._

I'm not personally a huge fan of the Streams API, but what you're referring to was overuse of the _Guava_ functional idioms, not Streams.

Reply:
@phrocker Many thanks for the link to the Guava wiki! The reminder that using streams (or functional programming constructs in general) should either reduce lines of code or provably improve performance has humbled me. I'll change my use of streams to @jrtom's imperative suggestion. :)